platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/ConfigurationInstaller.java

```diff
@@ -135,31 +171,38 @@ private void handleUpdate(ConfigurationEvent event) {
 
         FelixConfig felixConfig = new FelixConfig(configuration);
         String pid = felixConfig.getPid();
+
         File felixFileNew = felixConfig.getFelixFile();
         File felixFileOriginal = pidFileMap.get(pid);
 
-        // Config not tracked and will not be tracked? (1)
-        if (felixFileOriginal == null && felixFileNew == null) {
-            LOGGER.debug("Was not tracked and will not track pid {}", pid);
+        if (felixFileOriginal == null) {
+            // Config not tracked? (1)
+            if (felixFileNew == null) {
+                felixFileNew = felixConfig.generateDefaultFelixFile()
+                        .getFelixFile();
+            }
+
+            // We are seeing the config for the first time (2)
+            LOGGER.debug("Tracking pid {}", pid);
+            pidFileMap.put(pid, felixFileNew);
+            performOperation(configActions.create(felixFileNew.toPath(),
+                    felixConfig.getSanitizedProperties()));
             return;
         }
 
-        boolean filePropChanged = felixConfig.filePropChanged(felixFileOriginal);
-
-        // Felix file prop did not change? (4)
-        if (filePropChanged) {
-            // Are we seeing the config for the first time? (2) (3)
-            if (felixFileOriginal != null) {
-                try {
-                    felixConfig.setFelixFile(felixFileOriginal);
-                } catch (IOException e) {
-                    LOGGER.error("Could not set felix file name, error writing to config admin: ",
-                            e);
-                }
-                return;
-            }
-            pidFileMap.put(pid, felixFileNew);
-            LOGGER.debug("Tracking pid {}", pid);
+        // Felix file prop changed and we should revert it? (3)
+        if (felixConfig.filePropChanged(felixFileOriginal)) {
+            felixConfig.setFelixFile(felixFileOriginal);
+        }
+
+        // Respond to config properties being updated (4)
+        // Updating properties on disk, if they haven't changed, causes an infinite loop
+        // with the felix DirectoryWatcher
+        Path path = felixFileOriginal.toPath();
+        Dictionary<String, Object> fileState = configActions.getProperties(path);
+        Dictionary<String, Object> configState = felixConfig.getSanitizedProperties();
+        if (!equalDictionaries(configState, fileState)) {
+            performOperation(configActions.update(path, configState, false));
```

Comment:
We definitely do not want to reload from disk every time. Do you think caching the whole map would increase memory requirement too much? If so maybe just caching a checksum but then we have to compute that checksum every time. So I guess it is memory vs processing. But since config objects should not change that often, maybe the checksum alone is good enough.